api/v1alpha3/azurecluster_conversion.go

```diff
@@ -58,6 +57,32 @@ func (src *AzureCluster) ConvertTo(dstRaw conversion.Hub) error { // nolint
 	dst.Spec.NetworkSpec.APIServerLB.FrontendIPsCount = restored.Spec.NetworkSpec.APIServerLB.FrontendIPsCount
 	dst.Spec.NetworkSpec.NodeOutboundLB = restored.Spec.NetworkSpec.NodeOutboundLB
 
+	// Here we manually restore outbound security rules. Since v1alpha3 only supports ingress ("Inbound") rules, all v1alpha4 outbound rules are dropped when an AzureCluster
+	// is converted to v1alpha3. We loop through all security group rules. For all previously existing outbound rules we restore the full rule. For inbound rules we restore the Direction field if the rule still exists.
+	for i, restoredSubnet := range restored.Spec.NetworkSpec.Subnets {
+		for j, dstSubnet := range dst.Spec.NetworkSpec.Subnets {
+			if dstSubnet.Name == restoredSubnet.Name {
+				var restoredOutboundRules []infrav1alpha4.SecurityRule
+				for k, restoredSecurityRule := range restoredSubnet.SecurityGroup.SecurityRules {
+					if restoredSecurityRule.Direction == infrav1alpha4.SecurityRuleDirectionInbound {
+						// For inbound rules, we add back the restored direction if the rule still exists.
+						for l, dstSecurityRule := range dstSubnet.SecurityGroup.SecurityRules {
+							if dstSecurityRule.Name == restoredSecurityRule.Name {
+								dst.Spec.NetworkSpec.Subnets[j].SecurityGroup.SecurityRules[l].Direction = restored.Spec.NetworkSpec.Subnets[i].SecurityGroup.SecurityRules[k].Direction
+								break
+							}
+						}
```

Comment:
Isn't this logic redundant since inbound rules will get auto converted (now that you've set direction as well in line 263)? IIUC, I think we can simplify this to just handle outbound rules:
```
if restoredSecurityRule.Direction == infrav1alpha4.SecurityRuleDirectionOutbound {
  restoredOutboundRules = append(restoredOutboundRules, restoredSecurityRule)
}
```

wdyt?

Reply:
You are right! I had actually thought of that earlier and thought it would break the Fuzzy test because the directions set in the fuzzer are random strings. But, what I ended up doing is:

```
if restoredSecurityRule.Direction != infrav1alpha4.SecurityRuleDirectionInbound {
  restoredOutboundRules = append(restoredOutboundRules, restoredSecurityRule)
}
```

Which makes the Fuzzer happy, and, IMO, is more logically correct because the condition for "dropping" a rule in v1alpha4 --> v1alpha3 conversion is "if direction != Inbound" (we only include rules where direction == Ibound). So it makes sense to restore all rules that aren't inbound since they would all have been dropped theoretically speaking. And in the unlikely event that we need to add a new direction in v1alpha4 (for example, "both") we don't need to change the existing conversion.

Reply:
+1 checking for `!= inbound` instead of `==outbound`, more future proof!